cocos/core/gfx/device-manager.ts

```diff
@@ -0,0 +1,192 @@
+/*
+ Copyright (c) 2022 Xiamen Yaji Software Co., Ltd.
+
+ https://www.cocos.com/
+
+ Permission is hereby granted, free of charge, to any person obtaining a copy
+ of this software and associated engine source code (the "Software"), a limited,
+ worldwide, royalty-free, non-assignable, revocable and non-exclusive license
+ to use Cocos Creator solely to develop games on your target platforms. You shall
+ not use Cocos Creator software for developing other software or tools that's
+ used for developing games. You are not granted to publish, distribute,
+ sublicense, and/or sell copies of Cocos Creator.
+
+ The software or tools in this License Agreement are licensed, not sold.
+ Xiamen Yaji Software Co., Ltd. reserves all rights not expressly granted to you.
+
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
+ */
+
+import { JSB } from 'internal:constants';
+import { legacyCC } from '../global-exports';
+import { error, getError } from '../platform/debug';
+import { sys } from '../platform/sys';
+import { BindingMappingInfo, DeviceInfo, SwapchainInfo } from './base/define';
+import { Device } from './base/device';
+import { Swapchain } from './base/swapchain';
+import { BrowserType } from '../../../pal/system-info/enum-type';
+import { screen } from '../platform/screen';
+import { Settings, settings } from '../settings';
+
+/**
+ * @en
+ * Sets the renderer type, only useful on web
+ *
+ * @zh
+ * 渲染模式。
+ * 设置渲染器类型，仅适用于 web 端：
+ */
+export enum LegacyRenderMode {
+    /**
+     * @en
+     * Automatically chosen by engine.
+     * @zh
+     * 通过引擎自动选择。
+     */
+    AUTO = 0,
+    /**
+     * @en
+     * Forced to use canvas renderer.
+     * @zh
+     * 强制使用 canvas 渲染。
+     */
+    CANVAS = 1,
+    /**
+     * @en
+     * Forced to use WebGL renderer, but this will be ignored on mobile browsers.
+     * @zh
+     * 强制使用 WebGL 渲染，但是在部分 Android 浏览器中这个选项会被忽略。
+     */
+    WEBGL = 2,
+    /**
+     * @en
+     * Use Headless Renderer, which is useful in test or server env, only for internal use by cocos team for now
+     * @zh
+     * 使用空渲染器，可以用于测试和服务器端环境，目前暂时用于 Cocos 内部测试使用。
+     */
+    HEADLESS = 3
+}
+
+export enum RenderType {
+    UNKNOWN = -1,
+    CANVAS = 0,
+    WEBGL = 1,
+    OPENGL = 2,
+    HEADLESS = 3,
+}
+
+export class DeviceManager {
```

Comment:
introduce device manager to control gfx device creation

Reply:
Native already has `DeviceManager`, and it is different from it.

Reply:
For now I have mark this class to internal, and will export it when we are ready